### PR TITLE
rgw_sal_motr: [CORTX-30178] Handling null version-id use case, Enabling versioning for Multipart APIs, Delete and Bulk Delete APIs

### PR DESCRIPTION
**Pull request by sachitanands**
Merging 'versioning-feature' branch into 'main'

**Problem statement:**
  Need versioning for Multipart APIs, Delete and Bulk Delete APIs, handle null version-id usecase and fix issues for GET and HEAD object API.
  
**Solution**:
Fixed following issues for unversioned/version enabled/version suspended bucket.
1. Simple object null version-id cases.
2. Multipart upload null version-id cases.
3. Fixed delete-object and delete-objects issues.
4. Get and head object issues.

Signed-off-by: Sachitanand Shelake <sachitanand.shelake@seagate.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sachitanands** [comment]
unit tested manually using awscli commands.

**Review by cdeshmukh — APPROVED** [review]

**Review by priyanka25081999 — APPROVED** [review]

**Review by shriya-deshmukh — APPROVED** [review]

**Comment by andriytk** [comment]
Looks like this PR is the culprit of #311 and #312 regressions.

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -2715,6 +3010,15 @@ int MotrAtomicWriter::complete(size_t accounted_size, const std::string& etag,
...
                                                            << dendl;
+          
+  if(!info.versioning_enabled()) // suspended and un-versioned cases
+  {
+    // if bkt-version is suspended/unversioned, then fetch & update previous null
+    // version entry instead of adding new null version entry.
+    int ret_rc;
+    ret_rc = obj.update_null_reference(dpp, ent);
+    ldpp_dout(dpp, 20) <<__func__<< ": update_null_reference rc : " << ret_rc << dendl;
+  }
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:3021]: Why do we need this separate null-version entry at all in addition to the normal entry which we add below (line 3049)?

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -2698,6 +2987,12 @@ int MotrAtomicWriter::complete(size_t accounted_size, const std::string& etag,
                     << " etag: " << etag << " user_data=" << user_data << dendl;
   if (user_data)
     ent.meta.user_data = *user_data;
+
+  if (!obj.get_key().have_instance()) {
+    // generate-version-id for null version.
+    obj.gen_rand_obj_instance_name();
+    ent.key.instance = "null";
+   }
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:2995]: Why is this needed?
